File: Lab2_13_2/Lab2_13_2_Python/PhoneCallProcessor.py
from PhoneCall import PhoneCall


class PhoneCallProcessor():

    @staticmethod
    def get_cost_of_phone_call(phone_call: PhoneCall) -> float:
        COST_OF_PHONE_CALL_BETWEEN_9_20 = 1.50
        COST_OF_PHONE_CALL_REGULAR = 0.90
        return PhoneCallProcessor.__count_minutes_on_highest_tariff(phone_call) * COST_OF_PHONE_CALL_BETWEEN_9_20 + PhoneCallProcessor.__count_minutes_on_lowest_tariff(phone_call) * COST_OF_PHONE_CALL_REGULAR

    @staticmethod
    def __count_minutes_on_highest_tariff(phone_call: PhoneCall) -> int:
        nine_AM_in_minutes = PhoneCallProcessor.__convert_in_minutes("9:00")
        eigth_PM_in_minutes = PhoneCallProcessor.__convert_in_minutes("20:00")
        start_of_phone_call_in_minutes = PhoneCallProcessor.__convert_in_minutes(
            phone_call.start_of_phone_call)
        end_of_phone_call_in_minutes = PhoneCallProcessor.__convert_in_minutes(
            phone_call.end_of_phone_call)

        if(start_of_phone_call_in_minutes < nine_AM_in_minutes):
            if(end_of_phone_call_in_minutes < nine_AM_in_minutes):
                return 0
            elif(end_of_phone_call_in_minutes >= nine_AM_in_minutes and end_of_phone_call_in_minutes <= eigth_PM_in_minutes):
                return end_of_phone_call_in_minutes - nine_AM_in_minutes
            else:
                return PhoneCallProcessor.__get_length_of_interval_in_minutes(nine_AM_in_minutes, eigth_PM_in_minutes)
        elif(start_of_phone_call_in_minutes >= nine_AM_in_minutes and start_of_phone_call_in_minutes <= eigth_PM_in_minutes):
            if(end_of_phone_call_in_minutes >= nine_AM_in_minutes and end_of_phone_call_in_minutes <= eigth_PM_in_minutes):
                return PhoneCallProcessor.__get_length_of_interval_in_minutes(start_of_phone_call_in_minutes, end_of_phone_call_in_minutes)
            else:
                return eigth_PM_in_minutes - start_of_phone_call_in_minutes
        else:
            return 0

    @staticmethod
    def __count_minutes_on_lowest_tariff(phone_call: PhoneCall) -> int:
        return PhoneCallProcessor.get_current_phone_call_length(phone_call) - PhoneCallProcessor.__count_minutes_on_highest_tariff(phone_call)

    @staticmethod
    def __get_length_of_interval_in_minutes(interval_start: int, interval_end: int) -> int:
        return 1440 - interval_start - (1440 - interval_end)

    @staticmethod
    def __convert_in_minutes(time: str) -> int:
        values = time.split(":")
        return int(values[0]) * 60 + int(values[1])

    @staticmethod
    def __is_amount_of_cost_non_negative(phone_call) -> bool:
        return PhoneCallProcessor.get_cost_of_phone_call(phone_call)

    @staticmethod
    def get_current_phone_call_length(phone_call: PhoneCall) -> int:
        return PhoneCallProcessor.__get_length_of_interval_in_minutes(
            PhoneCallProcessor.__convert_in_minutes(
                phone_call.start_of_phone_call),
            PhoneCallProcessor.__convert_in_minutes(phone_call.end_of_phone_call))

    @staticmethod
    def stringify(phone_call: PhoneCall) -> str:
        if(PhoneCallProcessor.__is_amount_of_cost_non_negative(phone_call)):
            return "\nPhone number: " + phone_call.phone_number + "; \nStart of Phone Call: " + phone_call.start_of_phone_call + "; \nEnd of Phone Call: " + phone_call.end_of_phone_call + "; \nCost of Phone Call: " + str(PhoneCallProcessor.get_cost_of_phone_call(phone_call)) + "; \nLength of Phone Call in Minutes: " + str(PhoneCallProcessor.get_current_phone_call_length(phone_call)) + ";\n"
        return "\nNot appropriate input data: Phone call length cannot be less than 0 minutes"

File: Lab2_13_2/Lab2_13_2/PhoneCallProcessor.cs
using System;

namespace Lab2_13_2
{

    public class PhoneCallProcessor
    {
        // Public static methods block

        public static string Stringify(PhoneCall phoneCall)
        {
            decimal PhoneCallCost = GetCostOfPhoneCall(phoneCall);
            int PhoneCallLength = GetPhoneCallLength(phoneCall);
            if (IsAmountOfCostNonNegative(phoneCall))
            {
                return "Phone number: " + phoneCall.PhoneNumber +
                "; \nStart of Phone Call: " + phoneCall.TimeOfStartOfPhoneCall +
                "; \nEnd of Phone Call: " + phoneCall.TimeOfEndOfPhoneCall +
                "; \nCost of Phone Call: " + PhoneCallCost +
                "; \nLength of Phone Call in Minutes: " + PhoneCallLength +
                ";";
            }
            else
            {
                return "Not appropriate input data: Phone call length cannot be less than 0 minutes";
            }
        }

        public static decimal GetCostOfPhoneCall(PhoneCall phoneCall)
        {

            const decimal COST_OF_PHONE_CALL_BETWEEN_9_20 = 1.50M;
            const decimal COST_OF_PHONE_CALL_REGULAR = 0.90M;

            int MinutesOnHighestTariff = CountMinutesOnHighestTariff(phoneCall);
            int MinutesOnLowestTariff = CountMinutesOnLowestTariff(phoneCall);

            return MinutesOnHighestTariff * COST_OF_PHONE_CALL_BETWEEN_9_20 + MinutesOnLowestTariff * COST_OF_PHONE_CALL_REGULAR;
        }

        public static int GetPhoneCallLength(PhoneCall phoneCall)
        {
            int TimeOfStartOfPhoneCallInMinutes = ConvertInMinutes(phoneCall.TimeOfStartOfPhoneCall);
            int TimeOfEndOfPhoneCallInMinutes = ConvertInMinutes(phoneCall.TimeOfEndOfPhoneCall);

            return GetLengthOfPhoneCallInMinutes(TimeOfStartOfPhoneCallInMinutes, TimeOfEndOfPhoneCallInMinutes);
        }

        // Private static methods block

        private static int CountMinutesOnHighestTariff(PhoneCall phoneCall)
        {

            int NineAMInMinutes = ConvertInMinutes("9:00");
            int EigthPMInMinutes = ConvertInMinutes("20:00");
            int TimeOfStartOfPhoneCallInMinutes = ConvertInMinutes(phoneCall.TimeOfStartOfPhoneCall);
            int TimeOfEndOfPhoneCallInMinutes = ConvertInMinutes(phoneCall.TimeOfEndOfPhoneCall);
            int PhoneCallLength = GetPhoneCallLength(phoneCall);

            if (TimeOfStartOfPhoneCallInMinutes < NineAMInMinutes)
            {
                if (TimeOfEndOfPhoneCallInMinutes < NineAMInMinutes)
                {
                    return 0;
                }
                else if (TimeOfEndOfPhoneCallInMinutes >= NineAMInMinutes && TimeOfEndOfPhoneCallInMinutes <= EigthPMInMinutes)
                {
                    return TimeOfEndOfPhoneCallInMinutes - NineAMInMinutes;
                }
                else
                {
                    return GetLengthOfPhoneCallInMinutes(NineAMInMinutes, EigthPMInMinutes);
                }

            }
            else if (TimeOfStartOfPhoneCallInMinutes >= NineAMInMinutes && TimeOfStartOfPhoneCallInMinutes <= EigthPMInMinutes)
            {
                if (TimeOfEndOfPhoneCallInMinutes >= NineAMInMinutes && TimeOfEndOfPhoneCallInMinutes <= EigthPMInMinutes)
                {
                    return PhoneCallLength;
                }
                else
                {
                    return EigthPMInMinutes - TimeOfStartOfPhoneCallInMinutes;
                }
            }
            else
            {
                return 0;
            }
        }

        private static int CountMinutesOnLowestTariff(PhoneCall phoneCall)
        {
            int PhoneCallLength = GetPhoneCallLength(phoneCall);
            int MinutesOnHighestTariff = CountMinutesOnHighestTariff(phoneCall);
            return PhoneCallLength - MinutesOnHighestTariff;
        }

        private static int GetLengthOfPhoneCallInMinutes(int intervalStart, int intervalFinish)
        {
            return 1440 - intervalStart - (1440 - intervalFinish);
        }

        private static int ConvertInMinutes(string time)
        {
            string[] values = time.Split(":");
            return Convert.ToInt32(values[0]) * 60 + Convert.ToInt32(values[1]);
        }

        private static bool IsAmountOfCostNonNegative(PhoneCall phoneCall)
        {
            decimal PhoneCallCost = GetCostOfPhoneCall(phoneCall);

            return PhoneCallCost >= 0;
        }
    }
}
